glommio/src/sync/semaphore.rs

```diff
@@ -7,49 +7,115 @@ use std::cell::RefCell;
 use std::future::Future;
 use std::io::{Error, ErrorKind, Result};
 use std::pin::Pin;
-use std::rc::Rc;
 use std::task::{Context, Poll, Waker};
 
-use intrusive_collections::intrusive_adapter;
+use intrusive_collections::linked_list::LinkOps;
+use intrusive_collections::{container_of, offset_of, Adapter, PointerOps};
 use intrusive_collections::{LinkedList, LinkedListLink};
+use std::marker::PhantomPinned;
+use std::ptr::NonNull;
 
 #[derive(Debug)]
-struct Waiter {
-    node: Rc<WaiterNode>,
-    sem_state: Rc<RefCell<SemaphoreState>>,
+struct Waiter<'a> {
+    node: WaiterNode,
+    semaphore: &'a Semaphore,
 }
 
 #[derive(Debug)]
 struct WaiterNode {
     link: LinkedListLink,
-    state: RefCell<WaiterState>,
+    units: u64,
+    waker: RefCell<Option<Waker>>,
+
+    //waiter node can not be Unpin so its pointer could be used inside of intrusive
+    //collection, it also can not outlive the container which is guaranteed by the
+    //Waiter lifetime bound to the Semaphore which is container of all Waiters.
+    _p: PhantomPinned,
 }
 
-#[derive(Debug)]
-struct WaiterState {
-    units: u64,
-    waker: Option<Waker>,
+struct WaiterPointerOps;
+
+unsafe impl PointerOps for WaiterPointerOps {
+    type Value = WaiterNode;
+    type Pointer = NonNull<WaiterNode>;
+
+    unsafe fn from_raw(&self, value: *const Self::Value) -> Self::Pointer {
+        NonNull::new_unchecked(value as *mut Self::Value)
+    }
+
+    fn into_raw(&self, ptr: Self::Pointer) -> *const Self::Value {
+        ptr.as_ptr() as *const Self::Value
+    }
 }
 
-intrusive_adapter!(WaiterAdapter = Rc<WaiterNode> : WaiterNode {link : LinkedListLink});
+struct WaiterAdapter {
+    pointers_ops: WaiterPointerOps,
+    link_ops: LinkOps,
+}
+
+impl WaiterAdapter {
+    fn new() -> Self {
+        WaiterAdapter {
+            pointers_ops: WaiterPointerOps,
+            link_ops: LinkOps,
+        }
+    }
+}
+
+unsafe impl Adapter for WaiterAdapter {
+    type LinkOps = LinkOps;
+    type PointerOps = WaiterPointerOps;
+
+    unsafe fn get_value(
+        &self,
+        link: <Self::LinkOps as intrusive_collections::LinkOps>::LinkPtr,
+    ) -> *const <Self::PointerOps as PointerOps>::Value {
+        container_of!(link.as_ptr(), WaiterNode, link)
```

Comment:
can link be null? This doesn't seem to be a `NonNull`, at least it doesn't appear anywhere obvious

Reply:
Hi @glommer , those pointers can not be null because this trait implementation is used by intrusive_collections to convert pointers from one form to another (pointer to the link field to the pointer to the value field and vice versa). So those methods always receive valid pointers. Those pointers are not pointers to the NonNull container, those are pointers to the value contained into the collection and to the link inside of this value contained into collation.

Reply:
Also, that is a part of the contract that methods can not panic, and actually, they are useless with null pointers.

Reply:
Hi @glommer that is also NonNull object, so this LinkPtr can not be null.

Reply:
Do they document those things anywhere?
I am okay with that, but would be nice to point to some stated claims from the crate we are using.

Reply:
Sure that is part of the  trait code https://github.com/Amanieu/intrusive-rs/blob/master/src/linked_list.rs#L149 . We iplement LinkOps (it is in above type) and their LinkPtr is as NonNull<Link>

Reply:
LinkOps are defined there https://github.com/DataDog/glommio/pull/228/files/46aa06bd3f1f9cf978d2fdf94645591c168ae0ff#diff-9063a13a75caca6f9b4f70da0df9795d578d990677ae6c5ec0f916a43181d6c6R66 as result they determine that LinkPtr is NonNull<Link>